ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Params.hs

```diff
@@ -93,12 +94,20 @@ ascFromBits w = ascFromDouble $ toEnum (fromEnum w) / (2 ^ B.finiteBitSize w)
 ascVal :: Asc -> Double
 ascVal (Asc x) = x
 
+-- | Provides an Asc with high probability to produce k slots in
+-- a window of size s. Therefore, we chose @Asc >= 3k/s@.
```

Comment:
Did you choose this as we have `s = 3k/f` on mainnet? I think it is fine to keep it like this for now (especially as the tests show that this usually results in $k$ slots active out of $s$ with the currently generated values for $k$ and $s$).

In a a bit more principled fashion, I guess one could set a target false positive rate $p$ (eg $p = 1/1000$) and then choose $\textrm{Asc}$ such that $F(k-1,s,\mathrm{Asc}) \le p$ (where $F$ is the cumulative binomial distribution function, eg from `statistics`). Doesn't seem too complicated, but also not strictly necessary for now.

Reply:
> Did you choose this as we have s = 3k/f on mainnet?

That's indeed the case. I'll add a note about the choice.

I'd say the choice of mechanism to compute $Asc$ depends on whether we do want to test with less dense chains, as $3k/s$ might be produce sparse chains with too little probability.

Maybe we should aim for two generators, or a generator parameterized by $Asc$ or $p$.

Reply:
> Maybe we should aim for two generators, or a generator parameterized by $Asc$ or $p$.

Yeah, this feels similar to the second point in https://github.com/input-output-hk/ouroboros-consensus/pull/477#discussion_r1400438813.

Reply:
Why do we want this behavior?

If we use it for the honest schema, then it's unlikely we'll generate "minimally dense" honest chains, which seems like a mistake.

If we use it for alternative schemas, then it's unlikely we'll generate sparse alternative chains. That seems less like a mistake, but it's also not obvious that we never want to test with those, is it?

Reply:
> Why do we want this behavior?

From the PR description:

> Having more than k blocks after intersections makes the inputs interesting for the long range attack, and for testing that rollbacks of more than k can't happen in caught up nodes.

For long range attacks, it ensures that the node (at least if implemented correctly) did commit to a chain at the end of the test, which is what is tested for. I don't remember the exact reason motivation in the context of the rollback schedules, maybe @Niols can expand on that?

---

> If we use it for the honest schema, then it's unlikely we'll generate "minimally dense" honest chains, which seems like a mistake.

It is only used for alternative schemas, not for honest ones.

> If we use it for alternative schemas, then it's unlikely we'll generate sparse alternative chains. That seems less like a mistake, but it's also not obvious that we never want to test with those, is it?

Yeah, this is a good point, I think this might be indicative that parameterizing the generators over such desiderata might be useful (in order to make sure we can satisfy them when necessary, but still test with the full exhaustiveness in other cases), also see https://github.com/input-output-hk/ouroboros-consensus/pull/477#discussion_r1400438813.